mythril_core/src/ioapic.rs

```diff
@@ -0,0 +1,677 @@
+            gsi_base,
+        };
+
+        // From section 3.2.2 the version number of the I/O APIC
+        // should be 0x11
+        if ioapic.version() != IOAPIC_VERSION {
+            Err(Error::NotSupported)
+        } else {
+            Ok(ioapic)
+        }
+    }
+
+    /// Unsafe utility function for reading a 32-bit value from an
+    /// I/O APIC register.
+    ///
+    /// See section 3.0 of the I/O APIC specification.
+    unsafe fn read_raw(&self, reg: u8) -> u32 {
+        let addr = self.addr_lock.lock();
+        ptr::write_volatile(*addr, reg);
+        ptr::read_volatile(addr.offset(IOWIN_OFFSET) as *const u32)
+    }
+
+    /// Unsafe utility function for writing a 32-bit value to an
+    /// I/O APIC register.
+    ///
+    /// See section 3.0 of the I/O APIC specification.
+    unsafe fn write_raw(&self, reg: u8, val: u32) {
+        let addr = self.addr_lock.lock();
+        ptr::write_volatile(*addr, reg);
+        ptr::write_volatile(addr.offset(IOWIN_OFFSET) as *mut u32, val)
+    }
+
+    /// The ID of this I/O APIC.
+    ///
+    /// See section 3.2.1 of the I/O APIC specification.
+    pub fn id(&self) -> u8 {
+        let raw_id = unsafe { self.read_raw(0x00) };
+        ((raw_id >> 24) & 0x0f) as u8
+    }
+
+    /// Set the ID of this I/O APIC.
+    ///
+    /// See section 3.2.1 of the I/O APIC specification.
+    pub fn set_id(&self, id: u32) -> Result<()> {
+        if id > 0x0f {
+            Err(Error::InvalidValue(format!(
+                "I/O APIC ID `0x{:x}` too large",
+                id
+            )))
+        } else {
+            unsafe {
+                self.write_raw(0x00, id << 24);
+            }
+            Ok(())
+        }
+    }
+
+    /// The version of this I/O APIC.
+    ///
+    /// See section 3.2.2 of the I/O APIC specification.
+    pub fn version(&self) -> u8 {
+        let raw_version = unsafe { self.read_raw(0x01) };
+        (raw_version & 0xff) as u8
+    }
+
+    /// The maximum redirection entry for this I/O APIC.
+    ///
+    /// See section 3.2.2 of the I/O APIC specification.
+    pub fn max_redirection_entry(&self) -> u8 {
+        let raw_version = unsafe { self.read_raw(0x01) };
+        ((raw_version >> 16) & 0xff) as u8
+    }
+
+    /// The arbitration ID for this I/O APIC.
+    ///
+    /// See section 3.2.3 of the I/O APIC specification.
+    pub fn arbitration_id(&self) -> u8 {
+        let raw_arb_id = unsafe { self.read_raw(0x02) };
+        ((raw_arb_id >> 24) & 0x0f) as u8
+    }
+
+    /// Set the arbitration ID for this I/O APIC.
+    ///
+    /// See section 3.2.3 of the I/O APIC specification.
+    pub fn set_arbitration_id(&self, id: u8) -> Result<()> {
+        if id > 15 {
+            Err(Error::InvalidValue(format!(
+                "I/O APIC Arbitration ID `0x{:x}` too large",
+                id
+            )))
+        } else {
+            unsafe {
+                self.write_raw(0x02, (id as u32) << 24);
```

Comment:
What is the significance of 0x02 here? Should this be an enum?

Reply:
Ah, yeah it is the ID register. We should have constants for the I/O APIC Register offsets.